addons/misra.py

```diff
@@ -1579,6 +1580,49 @@ def misra_17_1(self, data):
             elif token.str == 'va_list':
                 self.reportError(token, 17, 1)
 
+    def misra_17_2(self, data):
+        def find_indirect_calls(func_id, calls_map, call_tokens_map, found_calls, visited):
```

Comment:
Using this function at [this for loop](https://github.com/danmar/cppcheck/pull/1962/files#diff-1bfaaa2a6da0266311b76e7291e15236R1619) I found all indirect recursive calls for function with a single `find_indirect_calls()` call and report all possible errors right on this cycle. If I'll looking for only first call, I should also consider a case when we have other indirect recursive calls for this function. So, I'll need to retry recursive search operation with additional `ignore` list parameters:

```python
        ignore = []
        for func_id in function_calls:
            passed = False
            while not passed:
                found = find_recursive_call(func_id, copy.deepcopy(function_calls), ignore)
                if found:
                    tok = call_tokens_map.get(tok_id)
                    ignore.append(found)
                    self.reportError(tok, 17, 2)
                 else:
                    passed = True
                    ignore = []
```

Reply:
I don't think it should be rewritten.. I just think the `found_calls` can be replaced with a bool. I removed the `found_calls` and removed the deepcopy stuff:
```
def find_recursive_call(search_for_function, call, calls_map, visited=set()):
        for tok in calls_map.get(call, []):
            if tok.function == search_for_function:
                return True
            if tok.function in visited:
                # This has already been handled
                continue
            visited.add(tok.function)
            if find_recursive_call(search_for_function, tok.function, calls_map, visited):
                return True
        return False
```
.. this code is completely untested so maybe it's off a little but well I hope you get the idea what I tried to do.

If this will be ineffective then could you please show some measurement?

Reply:
> I just think the found_calls can be replaced with a bool.

Consider the following source:
```c
void misra_17_2_a(void) {
  b(); // 17.2
}
void b(void) {
  a(); // 17.2
  a(); // 17.2
}
```

Where we should report an error? Using boolean return code, we'll report only at first `a()` call when search for `func_id == a.Id`. So I suggest to continue a search and check any other function calls.

Reply:
> If this will be ineffective then could you please show some measurement?

Of course not, when we break search at first occurrence, it will be faster.

It's easy to check with wrapping [this call](https://github.com/danmar/cppcheck/pull/1962/files#diff-1bfaaa2a6da0266311b76e7291e15236R2347):

```python
            import time
            start = time.time()
            self.misra_17_2(cfg)
            end = time.time()
            print(end - start)
```

Using dump from misra-test.c, I got the following result for current solution: `~0.0115058422089`.

Reply:
Well.. that is true. So if we determine that "calling a() inside b() will lead to recursion" then it would not be difficult to loop through b() and warn about all a() calls.

I think that there is complexity here that can be avoided:
 * deepcopy
 * 2 maps `function_calls` and `call_tokens_map`
 * Usage of "Id".

Reply:
I'm not sure how to simplify this.

> 2 maps function_calls and call_tokens_map. The "combined usage" is a bit hard to follow I think.
> Usage of "Id".

This required to avoid recursive copy for lxml tree elements using `deepcopy`.

> deepcopy

Required to modify elements between recursion cycles. The problem is that Python's dicts are mutable, and I must explicitly say that I use copy of `function_calls`. Another solution is add additional variable in `find_indirect_calls` to save the passed elements, but it is more cumbersome, I suppose.